File: java/src/org/smc/inputmethod/indic/settings/SettingsFragment.java
/*
 * Copyright (C) 2008 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.smc.inputmethod.indic.settings;

import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings.Secure;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.ViewGroup;

import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceScreen;
import androidx.recyclerview.widget.RecyclerView;

import com.android.inputmethod.latin.R;
import com.android.inputmethod.latin.utils.FeedbackUtils;

public final class SettingsFragment extends PreferenceFragmentCompat {
    // We don't care about menu grouping.
    private static final int NO_MENU_GROUP = Menu.NONE;
    // The first menu item id and order.
    private static final int MENU_ABOUT = Menu.FIRST;
    // The second menu item id and order.
    private static final int MENU_HELP_AND_FEEDBACK = Menu.FIRST + 1;

    @Override
    public void onCreatePreferences(final Bundle savedInstanceState, final String rootKey) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            getPreferenceManager().setStorageDeviceProtected();
        }
        addPreferencesFromResource(R.xml.prefs);
        addLanguageSelectionPreference();
    }

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    protected RecyclerView.Adapter onCreateAdapter(final PreferenceScreen preferenceScreen) {
        return new CardedPreferenceGroupAdapter(preferenceScreen);
    }

    @Override
    public RecyclerView onCreateRecyclerView(final LayoutInflater inflater, final ViewGroup parent,
            final Bundle savedInstanceState) {
        final RecyclerView recyclerView =
                super.onCreateRecyclerView(inflater, parent, savedInstanceState);
        recyclerView.addItemDecoration(new CardedPreferenceGroupAdapter.CardDivider(getActivity()));
        return recyclerView;
    }

    /**
     * Adds a "Languages" entry at the top that opens our self-managed language-selection screen.
     */
    private void addLanguageSelectionPreference() {
        final Activity activity = getActivity();
        final Preference languagePref = new Preference(activity);
        languagePref.setTitle(R.string.language_selection_title);
        languagePref.setIcon(R.drawable.ic_settings_languages);
        languagePref.setOrder(-1);
        languagePref.setFragment(LanguageSettingsFragment.class.getName());
        getPreferenceScreen().addPreference(languagePref);
    }

    @Override
    public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
        if (FeedbackUtils.isHelpAndFeedbackFormSupported()) {
            menu.add(NO_MENU_GROUP, MENU_HELP_AND_FEEDBACK /* itemId */,
                    MENU_HELP_AND_FEEDBACK /* order */, R.string.help_and_feedback);
        }
        final int aboutResId = FeedbackUtils.getAboutKeyboardTitleResId();
        if (aboutResId != 0) {
            menu.add(NO_MENU_GROUP, MENU_ABOUT /* itemId */, MENU_ABOUT /* order */, aboutResId);
        }
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        final Activity activity = getActivity();
        if (!isUserSetupComplete(activity)) {
            // If setup is not complete, it's not safe to launch Help or other activities
            // because they might go to the Play Store.  See b/19866981.
            return true;
        }
        final int itemId = item.getItemId();
        if (itemId == MENU_HELP_AND_FEEDBACK) {
            FeedbackUtils.showHelpAndFeedbackForm(activity);
            return true;
        }
        if (itemId == MENU_ABOUT) {
            final Intent aboutIntent = FeedbackUtils.getAboutKeyboardIntent(activity);
            if (aboutIntent != null) {
                startActivity(aboutIntent);
                return true;
            }
        }
        return super.onOptionsItemSelected(item);
    }

    private static boolean isUserSetupComplete(final Activity activity) {
        return Secure.getInt(activity.getContentResolver(), "user_setup_complete", 0) != 0;
    }
}

File: java/src/org/smc/inputmethod/indic/settings/SettingsFragment.kt
/*
 * Copyright 2026, Jishnu Mohan <jishnu@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.smc.inputmethod.indic.settings

import android.app.Activity
import android.os.Build
import android.os.Bundle
import android.provider.Settings.Secure
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.ViewGroup

import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceScreen
import androidx.recyclerview.widget.RecyclerView

import com.android.inputmethod.latin.R
import com.android.inputmethod.latin.utils.FeedbackUtils

@Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
class SettingsFragment : PreferenceFragmentCompat() {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            preferenceManager.setStorageDeviceProtected()
        }
        addPreferencesFromResource(R.xml.prefs)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateAdapter(preferenceScreen: PreferenceScreen): RecyclerView.Adapter<*> =
        CardedPreferenceGroupAdapter(preferenceScreen)

    override fun onCreateRecyclerView(
        inflater: LayoutInflater, parent: ViewGroup, savedInstanceState: Bundle?
    ): RecyclerView {
        val recyclerView = super.onCreateRecyclerView(inflater, parent, savedInstanceState)
        recyclerView.addItemDecoration(CardedPreferenceGroupAdapter.CardDivider(requireActivity()))
        recyclerView.isVerticalScrollBarEnabled = false
        return recyclerView
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        if (FeedbackUtils.isHelpAndFeedbackFormSupported()) {
            menu.add(NO_MENU_GROUP, MENU_HELP_AND_FEEDBACK, MENU_HELP_AND_FEEDBACK, R.string.help_and_feedback)
        }
        val aboutResId = FeedbackUtils.getAboutKeyboardTitleResId()
        if (aboutResId != 0) {
            menu.add(NO_MENU_GROUP, MENU_ABOUT, MENU_ABOUT, aboutResId)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val activity = requireActivity()
        if (!isUserSetupComplete(activity)) {
            // If setup is not complete, it's not safe to launch Help or other activities because
            // they might go to the Play Store. See b/19866981.
            return true
        }
        when (item.itemId) {
            MENU_HELP_AND_FEEDBACK -> {
                FeedbackUtils.showHelpAndFeedbackForm(activity)
                return true
            }
            MENU_ABOUT -> {
                val aboutIntent = FeedbackUtils.getAboutKeyboardIntent(activity)
                if (aboutIntent != null) {
                    startActivity(aboutIntent)
                    return true
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        private val NO_MENU_GROUP = Menu.NONE
        private val MENU_ABOUT = Menu.FIRST
        private val MENU_HELP_AND_FEEDBACK = Menu.FIRST + 1

        private fun isUserSetupComplete(activity: Activity): Boolean =
            Secure.getInt(activity.contentResolver, "user_setup_complete", 0) != 0
    }
}
